store/streaming/file/service.go

```diff
@@ -0,0 +1,279 @@
+		return err
+	}
+	// write all state changes cached for this stage to file
+	fss.stateCacheLock.Lock()
+	for _, stateChange := range fss.stateCache {
+		if _, err = dstFile.Write(stateChange); err != nil {
+			fss.stateCache = nil
+			fss.stateCacheLock.Unlock()
+			return err
+		}
+	}
+	// reset cache
+	fss.stateCache = nil
+	fss.stateCacheLock.Unlock()
+	// write res to file
+	lengthPrefixedResBytes, err := fss.codec.MarshalLengthPrefixed(&res)
+	if err != nil {
+		return err
+	}
+	if _, err = dstFile.Write(lengthPrefixedResBytes); err != nil {
+		return err
+	}
+	// close file
+	return dstFile.Close()
+}
+
+func (fss *StreamingService) openDeliverTxFile() (*os.File, error) {
+	fileName := fmt.Sprintf("block-%d-tx-%d", fss.currentBlockNumber, fss.currentTxIndex)
+	if fss.filePrefix != "" {
+		fileName = fmt.Sprintf("%s-%s", fss.filePrefix, fileName)
+	}
+	fss.currentTxIndex++
+	return os.OpenFile(filepath.Join(fss.writeDir, fileName), os.O_CREATE|os.O_WRONLY, 0600)
+}
+
+// ListenEndBlock satisfies the Hook interface
+// It writes out the received EndBlock request and response and the resulting state changes out to a file as described
+// in the above the naming schema
+func (fss *StreamingService) ListenEndBlock(ctx sdk.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error {
+	// generate the new file
+	dstFile, err := fss.openEndBlockFile()
+	if err != nil {
+		return err
+	}
+	// write req to file
+	lengthPrefixedReqBytes, err := fss.codec.MarshalLengthPrefixed(&req)
+	if err != nil {
+		return err
+	}
+	if _, err = dstFile.Write(lengthPrefixedReqBytes); err != nil {
+		return err
+	}
+	// write all state changes cached for this stage to file
+	fss.stateCacheLock.Lock()
+	for _, stateChange := range fss.stateCache {
+		if _, err = dstFile.Write(stateChange); err != nil {
+			fss.stateCache = nil
+			fss.stateCacheLock.Unlock()
+			return err
+		}
+	}
+	// reset cache
+	fss.stateCache = nil
+	fss.stateCacheLock.Unlock()
+	// write res to file
+	lengthPrefixedResBytes, err := fss.codec.MarshalLengthPrefixed(&res)
+	if err != nil {
+		return err
+	}
+	if _, err = dstFile.Write(lengthPrefixedResBytes); err != nil {
+		return err
+	}
+	// close file
+	return dstFile.Close()
+}
+
+func (fss *StreamingService) openEndBlockFile() (*os.File, error) {
+	fileName := fmt.Sprintf("block-%d-end", fss.currentBlockNumber)
+	if fss.filePrefix != "" {
+		fileName = fmt.Sprintf("%s-%s", fss.filePrefix, fileName)
+	}
+	return os.OpenFile(filepath.Join(fss.writeDir, fileName), os.O_CREATE|os.O_WRONLY, 0600)
+}
+
+// Stream spins up a goroutine select loop which awaits length-prefixed binary encoded KV pairs and caches them in the order they were received
+// Do we need this and an intermediate writer? We could just write directly to the buffer on calls to Write
+// But then we don't support a Stream interface, which could be needed for other types of streamers
+func (fss *StreamingService) Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) {
```

Comment:
Who passes a `quitChan` to this and when is it called? Would it be more straightfoward to just have a `Close` method on `StreamingService`?

Reply:
It would be the App passing the `quitChan` to synchronize shutdown on its shutdown. And It would definitely be, cleaner to have a `Close` method. In the plugin PR I've switched it to that, but I can go ahead and switch it over here.

Reply:
Actually, if it is OK, I propose we leave it using a `quitChan` for now as it will require a more significant refactor to switch to `Close` than I anticipated due to how `LoadStreamingServices` is currently written.